src/main/java/uk/ac/cam/cl/dtg/segue/comm/MailGunEmailManager.java

```diff
@@ -0,0 +1,180 @@
+                                           final EmailType emailType, final IsaacMailGunTemplate templateType,
+                                           @Nullable final Map<String, Object> templateVariablesOrNull,
+                                           @Nullable final Map<Long, Map<String, Object>> userVariablesOrNull)
+            throws FeignException {
+
+        // Lazily construct the MailGun messages API
+        this.createMessagesApiIfNeeded();
+
+        // Ensure all relevant email variable maps are not null
+        Map<Long, Map<String, Object>> userVariables = null != userVariablesOrNull ? userVariablesOrNull : Map.of();
+        Map<String, Object> templateVariables = new HashMap<>(null != templateVariablesOrNull ? templateVariablesOrNull : Map.of());
+        // Combine local and global template variables (non-recipient ones)
+        if (null != this.globalStringTokens) {
+            templateVariables.putAll(this.globalStringTokens);
+        }
+
+        // This expression:
+        //  - filters users by their email preferences and email verification status
+        //  - for each user U, extracts the name and email of U, and adds them to the (possibly empty) map of
+        //    `userVariables.get(U.getId())`
+        //  - creates a new map with emails as keys and the updated `userVariables` maps as values
+        Map<String, Object> recipientVariables = userDTOs.stream()
+                .filter(user -> {
+                    // FIXME - should system emails ignore this setting to avoid abuse?
+                    if (user.getEmailVerificationStatus() == EmailVerificationStatus.DELIVERY_FAILED) {
+                        return false;
+                    }
+                    try {
+                        UserPreference preference = userPreferenceManager.getUserPreference(
+                                SegueUserPreferences.EMAIL_PREFERENCE.name(), emailType.name(), user.getId());
+                        // If no preference is present, do not send email to this user.
+                        return preference != null && preference.getPreferenceValue();
+                    } catch (SegueDatabaseException e) {
+                        // Assume the worst and do not send the email to this user in the case of a DB issue
+                    }
+                    return false;
+                }).collect(
+                    Collectors.toMap(
+                        RegisteredUserDTO::getEmail,
+                        u -> Stream.of(
+                                ImmutableMap.of(
+                                "email", u.getEmail(),
+                                "givenName", u.getGivenName()
+                                ),
+                                userVariables.getOrDefault(u.getId(), ImmutableMap.of())
+                        ).flatMap(m -> m.entrySet().stream())
+                        .collect(Collectors.toMap(Entry::getKey, Entry::getValue, (e1, e2) -> e1))
```

Comment:
This is not at all clear. I really can't follow what it is trying to do overall, nor in the details. Why does the merge function just discard the second argument, for example?

Reply:
There is a comment at the top of this stream expression that explains most of it, but I'll add some more fine-grained comments to explain the details (once I work out what the hell I was doing here 😅)